proguard/src/main/java/com/flurry/proguard/UploadProguardMapping.java

```diff
@@ -0,0 +1,216 @@
+        if (response.getStatusLine().getStatusCode() != HttpURLConnection.HTTP_CREATED) {
+            throw new IOException("Failed to create Upload with Status Code: " +
+                    response.getStatusLine().getStatusCode() + " and Status Reason: "
+                    + response.getStatusLine().getReasonPhrase());
+        }
+
+        BufferedReader in = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));
+        String input;
+        StringBuilder resp = new StringBuilder();
+        while ((input = in.readLine()) != null) {
+            resp.append(input);
+        }
+        in.close();
+        JSONObject jsonObject = new JSONObject(resp.toString());
+        return jsonObject.getJSONObject("data").get("id").toString();
+    }
+
+    private void sendToUploadService(File file, String projectId, String uploadId, String token)
+            throws IOException {
+        String uploadServiceUrl = UPLOAD_BASE + projectId + "/" + uploadId;
+        HttpPost postRequest = new HttpPost(uploadServiceUrl);
+        Map<String, String> requestHeaders = getHeadersForUploadService(file.length(), token);
+        for (String key : requestHeaders.keySet()) {
+            postRequest.setHeader(key, requestHeaders.get(key));
+        }
+        postRequest.setEntity(new FileEntity(file));
+        HttpResponse response = httpClient.execute(postRequest);
+        int responseStatusCode = response.getStatusLine().getStatusCode();
+        if (responseStatusCode != 201 && responseStatusCode != 202) {
+            throw new IOException("Upload Service POST failed with Status Code: " + responseStatusCode +
+                            " and Status Reason: " + response.getStatusLine().getReasonPhrase());
+        }
+    }
+
+    public String checkUploadStatus (String projectId, String uploadId, String token)
+            throws IOException, InterruptedException {
+        String queryUrl = METADATA_BASE + "project/" + projectId + "/uploads/" +
+                uploadId + "?fields[upload]=uploadStatus,failureReason";
+        HttpGet getRequest = new HttpGet(queryUrl);
+        Map<String, String> requestHeaders = getDefaultHeaders(token);
+        for (String key : requestHeaders.keySet()) {
+            getRequest.setHeader(key, requestHeaders.get(key));
+        }
+        HttpResponse response = httpClient.execute(getRequest);
+        if (response.getStatusLine().getStatusCode() != 200) {
+            throw new IOException("Checking Upload Status failed with Status Code: " +
+                    response.getStatusLine().getStatusCode() + " and Status Reason: " +
+                    response.getStatusLine().getReasonPhrase());
+        }
+        String uploadStatus =  getJsonFromEntity(response.getEntity()).getJSONObject("data")
+                .getJSONObject("attributes").getString("uploadStatus");
+        while (!uploadStatus.equals("COMPLETED") && !uploadStatus.equals("FAILED")) {
```

Comment:
You're not checking `getStatusCode` in this loop. You also might loop forever, we should bail after 10 minutes or so.